File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/job/meta/SubmitJobRequest.scala
package com.webank.eggroll.core.deepspeed.job.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.constant.StringConstants
import com.webank.eggroll.core.meta.Deepspeed

import scala.collection.JavaConverters._
import scala.language.implicitConversions

case class ResourceOptions(
                            timeoutSeconds: Int = 0,
                            resourceExhaustedStrategy: String
                          )

case class SubmitJobRequest(sessionId: String = StringConstants.EMPTY,
                            name: String = StringConstants.EMPTY,
                            jobType: String = StringConstants.EMPTY,
                            worldSize: Int = 0,
                            commandArguments: Array[String] = Array(),
                            environmentVariables: Map[String, String] = Map(),
                            files: Map[String, Array[Byte]] = Map.empty,
                            zippedFiles: Map[String, Array[Byte]] = Map.empty,
                            resourceOptions: ResourceOptions,
                            options: Map[String, String] = Map()
                           )

object SubmitJobRequest {
  implicit def serialize(src: SubmitJobRequest): Array[Byte] = {
    val builder = Deepspeed.SubmitJobRequest.newBuilder()
      .setSessionId(src.sessionId)
      .setName(src.name)
      .setJobType(src.jobType)
      .setWorldSize(src.worldSize)
      .addAllCommandArguments(src.commandArguments.toList.asJava)
      .putAllEnvironmentVariables(src.environmentVariables.asJava)
      .putAllFiles(src.files.mapValues(ByteString.copyFrom).asJava)
      .putAllZippedFiles(src.zippedFiles.mapValues(ByteString.copyFrom).asJava)
      .setResourceOptions(
        Deepspeed.ResourceOptions.newBuilder()
          .setTimeoutSeconds(src.resourceOptions.timeoutSeconds)
          .setResourceExhaustedStrategy(src.resourceOptions.resourceExhaustedStrategy)
          .build())
      .putAllOptions(src.options.asJava)
    builder.build().toByteArray
  }

  implicit def deserialize(byteString: ByteString): SubmitJobRequest = {
    val proto = Deepspeed.SubmitJobRequest.parseFrom(byteString)
    SubmitJobRequest(
      sessionId = proto.getSessionId,
      name = proto.getName,
      jobType = proto.getJobType,
      worldSize = proto.getWorldSize,
      commandArguments = proto.getCommandArgumentsList.asScala.toArray,
      environmentVariables = proto.getEnvironmentVariablesMap.asScala.toMap,
      files = proto.getFilesMap.asScala.mapValues(_.toByteArray).toMap,
      zippedFiles = proto.getZippedFilesMap.asScala.mapValues(_.toByteArray).toMap,
      resourceOptions = ResourceOptions(
        timeoutSeconds = proto.getResourceOptions.getTimeoutSeconds,
        resourceExhaustedStrategy = proto.getResourceOptions.getResourceExhaustedStrategy
      ),
      options = proto.getOptionsMap.asScala.toMap
    )
  }
}

File: jvm/core/src/main/java/com/eggroll/core/pojo/SubmitJobRequest.java
package com.eggroll.core.pojo;

import com.eggroll.core.constant.StringConstants;
import com.google.protobuf.ByteString;
import com.webank.eggroll.core.meta.Deepspeed;
import lombok.Data;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Data
public class SubmitJobRequest implements RpcMessage {
    Logger log = LoggerFactory.getLogger(SubmitJobRequest.class);

    private String sessionId = StringConstants.EMPTY;
    private String name = StringConstants.EMPTY;
    private String jobType = StringConstants.EMPTY;
    private Integer worldSize = 0;
    private List<String> commandArguments = new ArrayList<>();
    private Map<String, String> environmentVariables = new HashMap<>();
    private Map<String, byte[]> files;
    private Map<String, byte[]> zippedFiles;
    private ResourceOptions resourceOptions;
    private Map<String, String> options = new HashMap<>();

    public Deepspeed.SubmitJobRequest toProto() {
        Deepspeed.SubmitJobRequest.Builder builder = Deepspeed.SubmitJobRequest.newBuilder();
        builder.setSessionId(this.sessionId)
                .setName(this.name)
                .setJobType(this.jobType)
                .setWorldSize(this.worldSize)
                .addAllCommandArguments(this.commandArguments)
                .putAllEnvironmentVariables(this.environmentVariables)
                .setResourceOptions(this.resourceOptions.toProto())
                .putAllOptions(this.options);
        if (this.files != null) {
            this.files.forEach((k, v) -> {
                builder.putFiles(k, ByteString.copyFrom(v));
            });
        }

        if (this.zippedFiles != null) {
            this.zippedFiles.forEach((k, v) -> {
                builder.putZippedFiles(k, ByteString.copyFrom(v));
            });
        }
        return builder.build();
    }

    public static SubmitJobRequest fromProto(Deepspeed.SubmitJobRequest submitJobRequest) {
        SubmitJobRequest result = new SubmitJobRequest();
        result.deserialize(submitJobRequest.toByteArray());
        return result;
    }


    @Override
    public byte[] serialize() {
        return toProto().toByteArray();
    }

    @Override
    public void deserialize(byte[] data) {
        try {
            Deepspeed.SubmitJobRequest submitJobRequest = Deepspeed.SubmitJobRequest.parseFrom(data);
            this.sessionId = submitJobRequest.getSessionId();
            this.name = submitJobRequest.getName();
            this.jobType = submitJobRequest.getJobType();
            this.worldSize = submitJobRequest.getWorldSize();
            this.commandArguments = submitJobRequest.getCommandArgumentsList();
            this.environmentVariables = submitJobRequest.getEnvironmentVariablesMap();
            this.options = submitJobRequest.getOptionsMap();
            if (submitJobRequest.getFilesMap() != null) {
                this.files = new HashMap<>();
                submitJobRequest.getFilesMap().forEach((k, v) -> {
                    this.files.put(k, v.toByteArray());
                });
            }
            if (submitJobRequest.getZippedFilesMap() != null) {
                this.zippedFiles = new HashMap<>();
                submitJobRequest.getZippedFilesMap().forEach((k, v) -> {
                    this.zippedFiles.put(k, v.toByteArray());
                });
            }
            this.resourceOptions = ResourceOptions.fromProto(submitJobRequest.getResourceOptions());
        } catch (Exception e) {
            log.error("deserialize error : ", e);
        }

    }
}
